Appconfig endpoint

nickvergessen opened the pull request:
- Required for #1447 

@LukasReschke @MorrisJobke

nickvergessen committed:
Make AppConfig part of the public API

Signed-off-by: Joas Schilling <coding@schilljs.com>

MorrisJobke commented:
* [ ] @nickvergessen mumbled something about "I wanted to move this somehow in the provisioning API" ...

nickvergessen committed:
Switch to public API

Signed-off-by: Joas Schilling <coding@schilljs.com>

MorrisJobke commented:
@nickvergessen Fails ;)

```
PHP Fatal error:  Call to a member function getUpdater() on null in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 695
```

nickvergessen commented:
@icewind1991

icewind1991 commented:
I would prefer to return a promise from `OCP.AppConfig.*` instead of passing callbacks

nickvergessen commented:
Can you just take care of fixing the failing phpunit tests which I didn't touch?

We can discuss promises later, although I don't feel a lot like breaking the current js API.

icewind1991 commented:
> PHP Warning:  opendir(/drone/src/github.com/nextcloud/server/data-autotest): failed to open dir: No such file or directory in /drone/src/github.com/nextcloud/server/tests/lib/TestCase.php on line 300

Drone messing up?

MorrisJobke commented:
> Drone messing up?

I restarted it. Let's see

MorrisJobke commented:
Rebased to trigger CI

nickvergessen commented:
@icewind1991 
> PHP Fatal error: Call to a member function getUpdater() on null in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 695

why are file tests always failing. :(

nickvergessen commented:
I think I found the issue, will fix myself

nickvergessen committed:
Add unit tests

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen commented:
Rebased for #1822 and added php unit tests 

@icewind1991 do you want to do the promise stuff in here, or can we do it in a different PR? I have no idea what that is and how it works.

MorrisJobke commented on core/js/public/appconfig.js:
```diff
@@ -0,0 +1,113 @@
+			data: options.data || {},
+			success: options.success,
+			error: options.error
+		})
```
Semicolon

nickvergessen replied:
Fixed

MorrisJobke commented on core/js/public/appconfig.js:
```diff
@@ -0,0 +1,113 @@
+	 * @param {string} key
+	 * @param {Object} [options]
+	 * @param {function} [options.success]
+	 * @param {function} [options.error]
```
Proper documentation for the parameters would be nice

nickvergessen replied:
Well no documentation means nothing you should use?
Feel free to add something if you think it should be.

nickvergessen committed:
Fix missing semicolon

Signed-off-by: Joas Schilling <coding@schilljs.com>

MorrisJobke commented on apps/provisioning_api/appinfo/routes.php:
```diff
@@ -61,5 +56,11 @@
+		['name' => 'AppConfig#getKeys', 'url' => '/api/v1/config/apps/{app}', 'verb' => 'GET'],
+		['name' => 'AppConfig#getValue', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'GET'],
+		['name' => 'AppConfig#setValue', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'POST'],
+		['name' => 'AppConfig#deleteKey', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'DELETE'],
```
Do we really want to name this `/api/` again?

This then results in stuff like `http://localhost/server/ocs/v2.php/apps/provisioning_api/api/v1/config/apps/files_sharing/outgoing_server2server_share_enabled`

nickvergessen replied:
this is what all ocs routes use

MorrisJobke commented:
Beside the naming comment this is fine by me and works 👍

rullzer commented:
LGTM
